### ISSUE TITLE
Cannot read property 'updateMode' of null

### ISSUE THREAD

**Issue by akosyakov** [microsoft/monaco-editor#125]
I am getting the following error sometimes on a model language change:

```
TypeError: Cannot read property 'updateMode' of null
    at Cursor._onModelModeChanged (http://localhost:8888/ride/vs/editor/editor.main.js:51022:25)
    at http://localhost:8888/ride/vs/editor/editor.main.js:50928:23
    at safeInvoke1Arg (http://localhost:8888/ride/vs/editor/editor.main.js:21879:20)
    at Model.OrderGuaranteeEventEmitter._emitEvents (http://localhost:8888/ride/vs/editor/editor.main.js:21863:17)
    at Model.EventEmitter.emit (http://localhost:8888/ride/vs/editor/editor.main.js:21801:22)
    at Model.TextModelWithTokens._emitModelModeChangedEvent (http://localhost:8888/ride/vs/editor/editor.main.js:52687:22)
    at Model.TextModelWithTokens.setMode (http://localhost:8888/ride/vs/editor/editor.main.js:52449:26)
    at http://localhost:8888/ride/vs/editor/editor.main.js:52143:23
    at Object.notifySuccess [as _notify] (http://localhost:8888/ride/vs/editor/editor.main.js:7734:59)
    at Object.enter (http://localhost:8888/ride/vs/editor/editor.main.js:7416:30)
```

I can reproduce it, but cannot come up with the reproducible sample.

**Comment by alexandrudima** [comment]
Hm, I believe it is the Cursor._onModelModeChanged from here:

https://github.com/Microsoft/vscode/blob/955bfdda607de11e8f258dd81bce085f555e6148/src/vs/editor/common/controller/cursor.ts#L233

What sort of events do you listen to on your side? 

This would indicate that there was a mode change event, which before reaching the cursor, reached in some other code path (unknown at this time) that lead to the cursor being disposed, only to finally be dispatched on the cursor, by which time, the cursor is disposed and it throws... or something like that.

Is it that you're listening to some language/model related events and doing something "special" in the event handlers?

**Comment by akosyakov** [comment]
In our case depending on a language we should create a different URI, e.g. the extension part depends on a language. So on a mode change event we create a new editor model with a new uri and assign it to an editor. Could it be that the editor disposes the cursor because of a model change?

**Comment by alexandrudima** [comment]
Yes, when using `editor.setModel(newModel)` the editor will:
* dispose the previous view
* dispose the previous cursor
* create a new view
* create a new cursor
* it keeps the configuration

**Comment by akosyakov** [comment]
@alexandrudima so basically we cannot assign a new editor model inside a mode event listener? because it always will lead to this exception

**Comment by alexandrudima** [comment]
Well, the current bits lead to this exception due to event handlers ordering. The mode change event reaches you, you get the cursor to be disposed, and then it reaches the cursor, which now is already disposed. It is unexpected in the sense that the cursor doesn't have code to protect against it.

TL;DR you can workaround this by assigning the editor model async or I can write a test & protection code inside the cursor for this so-far-unexpected case.

On the other hand, I don't fully understand the tricks you do with the model URIs and why you would create a model with a new URI on a mode change event. Maybe you can elaborate on this?

**Comment by akosyakov** [comment]
We use monaco editor to open in memory document and an editor model should use file URI with an extension for a language selected by a user instead of in memory URI, otherwise we won't be able to provide tailored to the selected language services, unfortunately it is based on a URI extension, not on a language ID.

**Comment by akosyakov** [comment]
@alexandrudima but thx for hints, i have several ideas how to work it around

**Comment by alexandrudima** [comment]
The code has changed a lot in the meantime and I don't think an exception would be thrown in this case.
